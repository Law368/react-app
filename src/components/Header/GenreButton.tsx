import React, {MouseEventHandler} from 'react';
import {useDispatch} from 'react-redux';
import {searchMode} from '../../actionCreators/searchMode';
import {Search} from '../../enums/enum';

export default function GenreButton() {
    const dispatch = useDispatch();
    const handleActiveClass = (event: any) => {
        dispatch(searchMode(Search.Genre));
        event.preventDefault();
        const target = event.currentTarget;
        const buttonList = document.querySelectorAll('.search__modeButton');
        buttonList.forEach((button) => {
            if (
                button === target &&
                !button.classList.contains('search__modeButton--active')
            ) {
                return button.classList.add('search__modeButton--active');
            }
            return button.classList.remove('search__modeButton--active');
        });
    };
    return (
        <button className="search__modeButton" onClick={handleActiveClass}>
            Genre
        </button>
    );
}
