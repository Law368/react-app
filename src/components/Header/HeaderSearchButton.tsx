import React from 'react';
import {useDispatch} from 'react-redux';
import {Screen} from '../../enums/enum';
import {screenType} from '../../actionCreators/screenType';

export default function HeaderSearchButton() {
    const dispatch = useDispatch();
    const handleScreenChange = () => {
        dispatch(screenType(Screen.MoviesList));
    };

    return (
        <button className="header__searchButton" onClick={handleScreenChange}>
            search
        </button>
    );
}
