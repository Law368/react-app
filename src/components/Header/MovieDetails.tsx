import React from 'react';
import Logo from './Logo';
import {MovieData} from '../../interfaces/interfaces';

export default function MovieDetails(props: any) {
    return (
        <>
            <button className="header__searchButton">search</button>
            <div className="header__movieDetails">
                <div className="header__moviePoster">
                    <img
                        src={props.moviesData[1].poster_path}
                        alt="Header poster image"
                    />
                </div>
                <div className="header__movieInfo">
                    <div className="header__movieHeading">
                        <h2 className="header__movieTitle">
                            {props.moviesData[1].title}
                        </h2>
                        <div className="header__movieRating">
                            {props.moviesData[1].vote_average}
                        </div>
                    </div>
                    <div className="header__movieGenre">
                        {props.moviesData[1].genres[0]}
                    </div>
                    <div className="header__movieSpecifics">
                        <span className="header__movieRelease">
                            {props.moviesData[1].release_date}
                        </span>
                        <span className="header__movieDuration">
                            {props.moviesData[1].runtime} min
                        </span>
                    </div>
                    <div className="header__movieDescription">
                        <p>{props.moviesData[1].overview}</p>
                    </div>
                </div>
            </div>
        </>
    );
}
