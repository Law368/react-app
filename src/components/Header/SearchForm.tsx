import React from 'react';
import SearchResults from './searchResults';

export default function SearchForm() {
    return (
        <>
            <form action="">
                <label htmlFor="site-search" className="search__title">
                    Find your movie
                </label>
                <input
                    type="search"
                    id="site-search"
                    name="q"
                    placeholder="Your search input"
                    className="search__input"
                />
                <div className="search__mode">
                    <div className="search__modeDescription">
                        search bt
                        <button className="search__modeButton search__modeButton--active">
                            Title
                        </button>
                        <button className="search__modeButton">Genre</button>
                    </div>
                </div>
                <button>Search</button>
                <SearchResults />
            </form>
        </>
    );
}
