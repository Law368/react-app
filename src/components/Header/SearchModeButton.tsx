import React, {MouseEventHandler} from 'react';
import log from '../../helpers/log';

export default function SearchModeButton(props: {
    name: string;
    isActive: boolean;
    handleClick: Function;
}) {
    return (
        <button
            type="button"
            className={`search__modeButton ${
                props.isActive ? 'search__modeButton--active' : ''
            }`}
            onClick={props.handleClick(props.name)}
        >
            {props.name}
        </button>
    );
}
