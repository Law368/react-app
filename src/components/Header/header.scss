.header {
    background-color: rgb(49, 49, 49);
    padding-top: 20px;
    font-family: Arial, Helvetica, sans-serif;

    &__logoContainer {
        margin-bottom: 50px;
        padding-left: 100px;
    }
    &__searchFormContainer {
        padding-left: 100px;
        padding-right: 100px;
        background-color: #313131;
    }
    &__searchResultsContainer {
        background-color: rgb(197, 197, 197);
    }

    &__searchResults {
        display: flex;
        height: 50px;
        padding-left: 100px;
        padding-right: 70px;
        font-weight: bold;
        align-items: center;
        color: rgb(77, 77, 77);
    }

    &__sortBy {
        display: flex;
        margin-right: 0;
        margin-left: auto;

        & span {
            margin-right: 20px;
        }
    }

    &__numberOfFilmsFound {
    }

    &__sortOption {
        margin-right: 35px;
        cursor: pointer;
        &--active {
            color: rgb(214, 93, 93);
        }
    }

    &--movieDetails {
        background-color: rgb(49, 49, 49);
        padding-top: 20px;
        font-family: Arial, Helvetica, sans-serif;
        display: grid;
        grid-template-columns: 1fr 1fr;
        padding: 20px 30px;
        padding-bottom: 40px;
    }

    &__movieDetailsContainer {
        // grid-column: 1/3;
        background-color: #313131;
    }

    &__movieDetails {
        display: grid;
    }

    &__moviePosterContainer {
        grid-column: 1/2;
    }

    &__movieInfo {
        grid-column: 2/3;
        color: #fff;
    }

    &__searchButton {
        cursor: pointer;
        background-color: #fff;
        border: 0;
        outline: 0;
        border-radius: 5px;
        color: rgb(214, 93, 93);
        font-size: 13px;
        text-transform: uppercase;
        font-weight: bold;
        width: 110px;
        height: 30px;
        margin-left: auto;
        margin-right: 0;
    }

    &__movieHeading {
        display: flex;
        align-items: center;
    }

    &__movieTitle {
        color: rgb(214, 93, 93);
        font-weight: 900;
        margin-right: 15px;
    }

    &__movieRating {
        color: aquamarine;
        font-weight: 700;
        border: 1px solid aquamarine;
        border-radius: 50%;
        width: 45px;
        height: 45px;
        padding: 12px;
        text-align: center;
    }
    &__movieGenre {
        font-weight: 200;
        font-size: 14px;
        margin-bottom: 15px;
    }

    &__movieSpecifics {
        font-weight: bold;
        margin-bottom: 30px;
    }

    &__movieRelease {
        margin-right: 15px;
    }
    &__movieDescription {
        font-size: 15px;
    }
    &__moviePosterContainer {
        padding: 0px 30px;
    }
    &__moviePoster {
        width: 300px;
    }
}

.logo {
    color: red;
    display: inline-block;
}

.search__title {
    display: block;
    text-transform: uppercase;
    font-weight: bold;
    color: #fff;
    margin-bottom: 20px;
}

.search__input {
    width: 100%;
    background-color: rgb(41, 41, 41);
    border-width: 0px;
    border: none;
    border-bottom: 2px solid red;
    line-height: 40px;
    color: #fff;

    &:focus {
        outline: none !important;
    }
}

.search__mode {
    display: flex;
    margin-bottom: 30px;
}

.search__input {
    margin-bottom: 20px;
}

.search__modeDescription {
    text-transform: uppercase;
    color: #fff;
    font-weight: bold;
    font-size: 11px;
    & span {
        margin-right: 15px;
    }
}

.search__modeButton {
    background-color: rgb(97, 97, 97);
    color: #fff;
    text-transform: uppercase;
    font-size: 11px;
    font-weight: bold;
    border-width: 0px;
    border: none;
    padding: 5px 20px;
    cursor: pointer;
    margin-right: 15px;

    &--active {
        background-color: rgb(214, 93, 93);
    }
}

.search__button {
    margin-right: 0;
    text-transform: uppercase;
    margin-left: auto;
    background-color: rgb(214, 93, 93);
    font-size: 11px;
    font-weight: bold;
    border-width: 0px;
    border: none;
    color: #fff;
    padding: 7px 60px;
    cursor: pointer;
}
