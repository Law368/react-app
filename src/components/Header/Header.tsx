import React from 'react';
import {connect, ConnectedProps} from 'react-redux';
import Logo from './Logo';
import SearchForm from './SearchForm';
import {MovieData} from '../../interfaces/interfaces';
import SearchResults from './searchResults';
import {Screen} from '../../enums/enum';
import MovieDetails from './MovieDetails';

import {changeScreen} from '../../actions/changeScreen';

function Header(props: Props) {
    if (props.a === Screen.MoviesList) {
        return (
            <header className="header">
                <div className="header__logoContainer">
                    <Logo />
                </div>
                <SearchForm />
            </header>
        );
    }
    if (props.a === Screen.Movie) {
        return (
            <>
                <header className="header--movieDetails">
                    <div className="header__logoContainer">
                        <Logo />
                    </div>
                    <MovieDetails moviesData={props.moviesData} />
                </header>
                <SearchResults />
            </>
        );
    }
}

// TO retrive state from my store and to say that I want to these actions to be used for changing the screen
// TODO: прокидывать в компонент только необходимые ему свойства state.
const mapStateToProps = (state: {screen: Screen}) => ({
    a: state.screen,
});

// const mapDispatchToProps = (dispatch: any) => ({
//     // changeScreen: () => dispatch(changeScreen(Screen.MoviesList)),
//     // changeToMovie: () => dispatch(changeScreen(Screen.Movie)),
//     setScreenType: (sreeenType: Screen) => dispatch(changeScreen(sreeenType)),
// });

const connector = connect(mapStateToProps);
type PropsFromRedux = ConnectedProps<typeof connector>;
// export default connect(mapStateToProps, mapDispatchToProps)(Header);

interface Props extends PropsFromRedux {
    a: Screen;
    moviesData: MovieData[];
}

export default connector(Header);

// Connect,mapStateToProps, mapDispatchToProps и какие еще 2 параметра можно добавить в hoc Connect
