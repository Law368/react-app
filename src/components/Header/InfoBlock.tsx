import React from 'react';
import {connect, ConnectedProps, useSelector} from 'react-redux';
import {Header} from './Header';
import SearchForm from './SearchForm';
import {MovieData} from '../../interfaces/interfaces';
import SearchResultsContainer from './SearchResults/searchResultsContainer';
import {Screen, Search} from '../../enums/enum';
import MovieDetails from '../movieDetails/MovieDetails';
import HeaderSearchButton from '../movieDetails/HeaderSearchButton';
import {State} from '../../reducers/commonReducer';

function InfoBlock(props: Props) {
    const screenType = useSelector((state: State) => state.screen);
    if (props.screen === Screen.MoviesList) {
        return (
            <>
                <Header placement={'header'} />
                <SearchForm searchType={props.searchType} />
                <SearchResultsContainer movieList={props.moviesData} />
            </>
        );
    }
    if (props.screen === Screen.Movie) {
        return (
            <>
                <Header
                    placement={'header'}
                    displayMode={screenType === Screen.Movie}
                />
                <HeaderSearchButton />
                <MovieDetails moviesData={props.moviesData} />
                <SearchResultsContainer movieList={props.moviesData} />
            </>
        );
    }
}

const mapStateToProps = (state: {
    screen: Screen;
    searchMode: Search;
    currentMovie: MovieData;
}) => ({
    screen: state.screen,
    searchMode: state.searchMode,
});

const connector = connect(mapStateToProps);
type PropsFromRedux = ConnectedProps<typeof connector>;

export interface Props extends PropsFromRedux {
    screen: Screen;
    moviesData: MovieData[];
    searchType: Search;
}

export default connector(InfoBlock);
