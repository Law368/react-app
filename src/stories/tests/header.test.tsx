import React from 'react';
import renderer from 'react-test-renderer';

import {Header} from '../../components/Header/Header';

it("It shouldn't have a class header--movieDetails", () => {
    const element = renderer.create(<Header placement={'header'} />).toJSON();
    expect(element).toMatchSnapshot();
});

// TODO: 1. Убрать вызовы хуков вне функциональных компонентов (fetchMoviedataById, fetchMoviesBySearchInput) и вместо этого использовать thunk actionы и диспатчить их +++
// TODO: 2. Написать еще пару Unit тестов помимо сортировки. +++
// TODO: 3. Закончить Snapshot тесты (добавить тест на активность/неактивность кнопки)) +++
// TODO: 4. Посмотреть что такое End to End тесты выбрать библеотеку для них и попытаться написать один сценарий и запустить его +++
// TODO: 5. Переименовать папу movieDetails с большой буквы
// TODO: 6. Если что то не испоьзуется, то удалять из файлов
