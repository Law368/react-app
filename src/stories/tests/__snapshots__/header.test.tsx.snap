// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It shouldn't have a class header--movieDetails 1`] = `
<header
  className="header "
>
  <div
    className="header__logoContainer"
  >
    <span
      className="logo"
    >
      netflixroullete
    </span>
  </div>
</header>
`;
